pkg/virt-operator/kubevirt.go

```diff
@@ -422,6 +454,190 @@ func (c *KubeVirtController) execute(key string) error {
 	return syncError
 }
 
+func (c *KubeVirtController) generateInstallStrategyJob(kv *v1.KubeVirt, imageTag string, imageRegistry string) *batchv1.Job {
+
+	pullPolicy := k8sv1.PullIfNotPresent
+	if string(kv.Spec.ImagePullPolicy) != "" {
+		pullPolicy = kv.Spec.ImagePullPolicy
+	}
+	job := &batchv1.Job{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "batch/v1",
+			Kind:       "Job",
+		},
+
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: kv.Namespace,
+			Name:      fmt.Sprintf("virt-install-strategy-job-%s", imageTag),
+			Labels: map[string]string{
+				v1.AppLabel:                    "",
+				v1.ManagedByLabel:              v1.ManagedByLabelOperatorValue,
+				v1.InstallStrategyVersionLabel: imageTag,
+			},
+		},
+		Spec: batchv1.JobSpec{
+			Template: k8sv1.PodTemplateSpec{
+
+				Spec: k8sv1.PodSpec{
+					ServiceAccountName: "kubevirt-operator",
+					RestartPolicy:      k8sv1.RestartPolicyNever,
+					Containers: []k8sv1.Container{
+						{
+							Name:            "install-strategy-upload",
+							Image:           fmt.Sprintf("%s/%s:%s", imageRegistry, "virt-operator", imageTag),
+							ImagePullPolicy: pullPolicy,
+							Command: []string{
+								"virt-operator",
+								"--dump-install-strategy",
+							},
+							Env: []k8sv1.EnvVar{
+								{
+									Name:  util.OperatorImageEnvName,
+									Value: fmt.Sprintf("%s/%s:%s", imageRegistry, "virt-operator", imageTag),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return job
+}
+
+func (c *KubeVirtController) garbageCollectInstallStrategyJobs() error {
+	batch := c.clientset.BatchV1()
+	jobs := c.stores.InstallStrategyJobCache.List()
+
+	for _, obj := range jobs {
+		job, ok := obj.(*batchv1.Job)
+		if !ok {
+			continue
+		}
+		if job.Status.CompletionTime == nil {
+			continue
+		}
+
+		propagationPolicy := metav1.DeletePropagationForeground
+		err := batch.Jobs(job.Namespace).Delete(job.Name, &metav1.DeleteOptions{
+			PropagationPolicy: &propagationPolicy,
+		})
+		if err != nil {
+			return err
+		}
+		log.Log.Object(job).Errorf("Garbage collected completed install strategy job")
+	}
+
+	return nil
+}
+
+// Loads install strategies into memory, and generates jobs to
+// create install strategies that don't exist yet.
+func (c *KubeVirtController) loadInstallStrategy(kv *v1.KubeVirt) (*installstrategy.InstallStrategy, bool, error) {
+
+	kvkey, err := controller.KeyFunc(kv)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// 1. see if we already loaded the install strategy
+	c.installStrategyMutex.Lock()
+	strategy, ok := c.installStrategyMap[c.config.ImageTag]
+	// don't use a defer here because we don't want to block this lock
+	// while we make API calls later on in the function
+	c.installStrategyMutex.Unlock()
+
+	if ok {
+		// we already loaded this strategy into memory
```

Comment:
since we can't prevent that people use the latest tag, I'm wondering if it needs special handling, like always pulling a new image and recreate the installStrategy?

Reply:
oh good point. That's tricky.

maybe the operator should just clear all install strategy configMaps when the KubeVirt object is deleted.  That would mean we'd start fresh for every new install.

Reply:
sounds good to me

Reply:
fixed